testutils/testtracing/propagation.go

```diff
@@ -0,0 +1,313 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testtracing
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/testutils"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/mocktracer"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uber/jaeger-client-go"
+	"golang.org/x/net/context"
+)
+
+const (
+	// BaggageKey is used for testing baggage propagation
+	BaggageKey = "luggage"
+
+	// BaggageValue is used for testing baggage propagation
+	BaggageValue = "suitcase"
+)
+
+// TracingRequest tests tracing capabilities in a given server.
+type TracingRequest struct {
+	// ForwardCount tells the server how many times to forward this request to itself recursively
+	ForwardCount int
+}
+
+// TracingResponse captures the trace info observed in the server and its downstream calls
+type TracingResponse struct {
+	TraceID        uint64
+	SpanID         uint64
+	ParentID       uint64
+	TracingEnabled bool
+	Child          *TracingResponse
+	Luggage        string
+}
+
+// ObserveSpan extracts an OpenTracing span from the context and populates the response.
+func (r *TracingResponse) ObserveSpan(ctx context.Context) *TracingResponse {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		if mockSpan, ok := span.(*mocktracer.MockSpan); ok {
+			sc := mockSpan.Context().(*mocktracer.MockSpanContext)
+			r.TraceID = uint64(sc.TraceID)
+			r.SpanID = uint64(sc.SpanID)
+			r.ParentID = uint64(mockSpan.ParentID)
+			r.TracingEnabled = sc.Sampled
+		} else if span := tchannel.CurrentSpan(ctx); span != nil {
+			r.TraceID = span.TraceID()
+			r.SpanID = span.SpanID()
+			r.ParentID = span.ParentID()
+			r.TracingEnabled = span.Flags()&1 == 1
+		}
+		r.Luggage = span.Context().BaggageItem(BaggageKey)
+	}
+	return r
+}
+
+// TraceHandler is a base class for testing tracing propagation
+type TraceHandler struct {
+	Ch *tchannel.Channel
+}
+
+// HandleCall is used by handlers from different encodings as the main business logic.
+// It respects the ForwardCount input parameter to make downstream calls, and returns
+// a result containing the observed tracing span and the downstream results.
+func (h *TraceHandler) HandleCall(
+	ctx context.Context,
+	req *TracingRequest,
+	downstream TracingCall,
+) (*TracingResponse, error) {
+	var childResp *TracingResponse
+	if req.ForwardCount > 0 {
+		downstreamReq := &TracingRequest{ForwardCount: req.ForwardCount - 1}
+		if resp, err := downstream(ctx, downstreamReq); err == nil {
+			childResp = resp
+		} else {
+			return nil, err
+		}
+	}
+
+	resp := &TracingResponse{Child: childResp}
+	resp.ObserveSpan(ctx)
+
+	return resp, nil
+}
+
+// TracerType is a convenient enum to indicate which type of tracer is being used in the test
+type TracerType string
```

Comment:
is this ever printed? if not, consider just having a custom int type and using `iota`